From 968b52c7fea609ae7efcb94fd69a1bb55c22973a Mon Sep 17 00:00:00 2001
From: "Antoni S. Puimedon" <asegurap@redhat.com>
Date: Fri, 10 Oct 2014 13:04:30 +0200
Subject: [PATCH] nwfilter: Do not attempt to re-define the nwfilter when not
 undefining
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Undefining a filter is an operation that can fail. The fact that we
had the same exception swallowing for two operations, looking up and
discovering led us to the wrong assumption then of defining again the
filter even when it had been looked up fine but could not be
undefined.

The most sensible behavior to take for 3.5 is to just consider the
failure to undefine properly and re-define it only in the case of
failure to look up.

Change-Id: Ia284c0554ef86a4668808b26523fbb3b51baebcc
Bug-Url: https://bugzilla.redhat.com/1150718
Signed-off-by: Antoni S. Puimedon <asegurap@redhat.com>
Reviewed-on: http://gerrit.ovirt.org/33993
Reviewed-by: Dan Kenigsberg <danken@redhat.com>
Tested-by: Petr Horáček <phoracek@redhat.com>
---
 lib/vdsm/tool/nwfilter.py | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/lib/vdsm/tool/nwfilter.py b/lib/vdsm/tool/nwfilter.py
index 19c95f6..1b7537c 100755
--- a/lib/vdsm/tool/nwfilter.py
+++ b/lib/vdsm/tool/nwfilter.py
@@ -60,12 +60,18 @@ class NwFilter(object):
         """
 
         try:
-            conn.nwfilterLookupByName(self.filterName).undefine()
-        except libvirt.libvirtError:
-            # Ignore failure if filter isn't exists or if failed to remove.
-            # Failure might occur when attempting to remove a filter which
-            # is being used by running VMs
+            old_filter = conn.nwfilterLookupByName(self.filterName)
+        except libvirt.libvirtError:  # No filter, we can define it.
             pass
+        else:
+            try:
+                old_filter.undefine()
+            except libvirt.libvirtError:
+                # If the filter is in use it may fail to be removed. In that
+                # case we warn of it and consider the task done.
+                logging.warning('Failed to remove the old filter %s.',
+                                old_filter.name(), exc_info=True)
+                return
 
         nwFilter = conn.nwfilterDefineXML(self.buildFilterXml())
         logging.debug("Filter %s was defined", nwFilter.name())
-- 
1.8.3.1

